skel/opt/cartesi/bin/init

```diff
@@ -1,85 +1,80 @@
-busybox ifconfig lo 127.0.0.1
+[ -z "$noloopback" ] && busybox ifconfig lo up
 
-# cmdline application
-if [ -n "$*" ]; then
-    [ -f /etc/environment ] && \
-        source /etc/environment && \
-        export PATH=/opt/cartesi/bin:"$PATH" # put it back on PATH
+# source environment
+[ -f /etc/environment ] && . /etc/environment
+export PATH="$PATH:/opt/cartesi/bin"
 
-    # can login as dapp user?
-    if [ ! "$single" == "yes" ] && busybox id dapp &> /dev/null; then
-        if [ -c /dev/rollup ]; then
-            busybox chown root:dapp /dev/rollup
-            busybox chmod 660 /dev/rollup
-        fi
-        if [ -c /dev/yield ]; then
-            busybox chown root:dapp /dev/yield
-            busybox chmod 660 /dev/yield
-        fi
-        export HOME=~dapp USER=dapp LOGNAME=dapp
-    else
-        export HOME=~root USER=root LOGNAME=root
-    fi
+# execute init from device tree when available
+[ -f /proc/device-tree/cartesi-machine/init ] && . /proc/device-tree/cartesi-machine/init
 
-    # execute cmdline
-    cd $HOME &&
+# use entrypoint from device tree when available, otherwise from command line
+if [ -s /proc/device-tree/cartesi-machine/entrypoint ]; then
+    ENTRYPOINT=$(busybox cat /proc/device-tree/cartesi-machine/entrypoint)
+elif [ -n "$*" ]; then
+    ENTRYPOINT="$*"
+fi
+
+# is entrypoint not empty?
+if [ -n "$ENTRYPOINT" ]; then
+    USER=${USER:-root}
+    HOME=$(eval echo ~$USER)
+    WORKDIR=${WORKDIR:-"$HOME"}
+
+    # give user group access to rollup and yield devices
+    [ -c /dev/rollup ] && \
+        busybox chown :$(busybox id -g $USER) /dev/rollup && \
+        busybox chmod g+rw /dev/rollup
+    [ -c /dev/yield ] &&
+        busybox chown :$(busybox id -g $USER) /dev/yield && \
+        busybox chmod g+rw /dev/yield
+
+    # execute entrypoint
+    (cd $WORKDIR &&
+        USER=$USER LOGNAME=$USER HOME="$HOME" \
         busybox setsid \
         busybox cttyhack \
-        busybox su -p $USER -c "$*"
-    RC=$?
+        busybox su -p $USER -c "$ENTRYPOINT")
 else
     echo "Nothing to do."
-    RC=0
 fi
+RC=$?
 
+# shutdown
 busybox mount -o ro,remount /
 busybox umount -af
-[ $RC == 0 ] && busybox poweroff -f || xhalt "$RC"
+if [ $RC != 0 ] && [ -x /opt/cartesi/bin/xhalt ]; then
+    /opt/cartesi/bin/xhalt "$RC"
+fi
+busybox poweroff -f
+
+# Copyright Cartesi and individual authors (see AUTHORS)
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```

Comment:
If you are moving this to the end just so you save a few mcycles, at least add a terminating end-of-line to the file. How did you even create a file without the end-of-line? :smile: